backends/tc/tcam_program.cpp

```diff
@@ -0,0 +1,218 @@
+#include "backends/tc/tcam_program.h"
+
+#include <algorithm>
+#include <numeric>
+
+#include "absl/strings/match.h"
+#include "absl/strings/string_view.h"
+#include "backends/tc/util.h"
+
+namespace backends::tc {
+
+std::ostream& operator<<(std::ostream& out, const TCAMEntry& entry) {
+  out << "TCAMEntry { value = " << util::BitStringToString(entry.value)
+      << " mask = " << util::BitStringToString(entry.mask)
+      << " instructions = [ ";
+  for (const auto& instruction : entry.instructions) {
+    out << instruction->ToString() << "; ";
+  }
+  out << "] }";
+  return out;
+}
+
+size_t TCAMProgram::HeaderSize(absl::string_view header_name) const {
+  if (auto header_fields = HeaderFields(header_name)) {
+    size_t header_size = 0;
+    for (const auto& field : *header_fields) {
+      header_size += field.second;
+    }
+    return header_size;
+  } else {
+    // The header's type is unknown, assume it is zero
+    return 0;
+  }
+}
+
+const std::vector<std::pair<FieldPath, size_t>>* TCAMProgram::HeaderFields(
+    absl::string_view header_name) const {
+  auto header_type = header_instantiations_.find(header_name);
+  if (header_type == header_instantiations_.end()) {
+    // The header's type is unknown
+    return nullptr;
+  }
+  return &header_type_definitions_.at(header_type->second);
+}
+
+absl::optional<std::pair<size_t, size_t>> TCAMProgram::FieldOffsetAndSize(
+    const FieldPath& field_path) const {
+  const auto header_end = field_path.find('.');
+  const auto header = field_path.substr(0, header_end);
+  const auto& header_type = header_instantiations_.at(header);
+  if (header_end == std::string::npos) {
+    // The field_path refers to the whole header, the offset is 0, the size is
+    // the total size of the header
+    const auto& header_typedef = header_type_definitions_.at(header_type);
+    // When switching to C++17, the calculation of `header_size` can be
```

Comment:
I originally wrote the backend in C++17. I left this kind of comments when I was backporting it to C++11 and some language features or library functions were not available, so that the code-base can be cleaned up if/when p4c switches to C++17. I can make a pass to remove these comments if upgrading to a newer version of C++ is not on the horizon.